### ISSUE TITLE
NoMethodError: undefined method `[]' for nil:NilClass when querying with AR4

### ISSUE THREAD

**Issue by lowjoel** [aha-app/calculated_attributes#4]
Sorry, I tried to build a spec to demonstrate this but I don't know AR/Arel well enough to figure out what's going on.

Basically: in AR4.2.4, lib/active_record/associations/join_dependency.rb:259. `seen[parent.base_klass][primary_id][node.base_klass][id]`.

The compatible `seen` hash constructed in lib/rails_4_patches.rb:35.
```ruby
        seen = Hash.new do |i, object_id|
          i[object_id] = Hash.new do |j, child_class|
            j[child_class] = {}
          end
        end
```

I think this really ought to be a spec, but I don't know how to isolate a test case that will trigger this.

My proposed fix:
```ruby
        seen = Hash.new do |k, parent_class|
          k[parent_class] = Hash.new do |i, object_id|
            i[object_id] = Hash.new do |j, child_class|
              j[child_class] = {}
            end
          end
        end
```
The monkey-patching is getting a bit nasty now...

**Comment by lowjoel** [comment]
Is this an eager loading/joins problem? I'm trying all sorts of examples to get it to break in the gem's specs.

The query that fails is something like:
```sql
"SELECT "courses"."id" AS t0_r0, "courses"."instance_id" AS t0_r1, "courses"."title" AS t0_r2, "courses"."description" AS t0_r3, "courses"."logo" AS t0_r4, "courses"."status" AS t0_r5, "courses"."registration_key" AS t0_r6, "courses"."settings" AS t0_r7, "courses"."start_at" AS t0_r8, "courses"."end_at" AS t0_r9, "courses"."creator_id" AS t0_r10, "courses"."updater_id" AS t0_r11, "courses"."created_at" AS t0_r12, "courses"."updated_at" AS t0_r13, "course_users"."id" AS t1_r0, "course_users"."course_id" AS t1_r1, "course_users"."user_id" AS t1_r2, "course_users"."workflow_state" AS t1_r3, "course_users"."role" AS t1_r4, "course_users"."name" AS t1_r5, "course_users"."phantom" AS t1_r6, "course_users"."last_active_at" AS t1_r7, "course_users"."created_at" AS t1_r8, "course_users"."updated_at" AS t1_r9, "course_users"."creator_id" AS t1_r10, "course_users"."updater_id" AS t1_r11, "instances"."id" AS t2_r0, "instances"."name" AS t2_r1, "instances"."host" AS t2_r2, "instances"."settings" AS t2_r3, "instance_users"."id" AS t3_r0, "instance_users"."instance_id" AS t3_r1, "instance_users"."user_id" AS t3_r2, "instance_users"."role" AS t3_r3, "instance_users"."created_at" AS t3_r4, "instance_users"."updated_at" AS t3_r5 FROM "courses" LEFT OUTER JOIN "course_users" ON "course_users"."course_id" = "courses"."id" LEFT OUTER JOIN "instances" ON "instances"."id" = "courses"."instance_id" LEFT OUTER JOIN "instance_users" ON "instance_users"."instance_id" = "instances"."id" AND "instance_users"."instance_id" = 1618 WHERE "courses"."instance_id" = 1618 AND ('t'='t')"
```

It's generated by cancancan, so it's a little hard for me to isolate a repro. I know that the column aliasing takes place only when joins/includes are used...

**Comment by lowjoel** [comment]
Got it!

see lowjoel/calculated_attributes@d96bb7ec85ad86415a5bc4d9aac75e8e1235f2cb

**Comment by schneidmaster** [comment]
Thanks a lot. I'll see if I can trace this down.

**Comment by lowjoel** [comment]
Let me know if there's anything else you need. It'd be great if this is fixed :)

**Comment by schneidmaster** [comment]
Closed in #6 - thanks for the PR.
